File: bindings/elixir/lib/umicp/matrix/operations.ex
defmodule UMICP.Matrix.Operations do
  @moduledoc """
  Matrix and vector operations for UMICP.
  
  Provides high-performance matrix operations using either Nx (Numerical Elixir)
  when available, or pure Elixir fallback implementations.
  """

  @doc """
  Add two vectors.

  ## Examples

      iex> UMICP.Matrix.Operations.vector_add([1.0, 2.0], [3.0, 4.0])
      {:ok, [4.0, 6.0]}
  """
  @spec vector_add([float()], [float()]) :: {:ok, [float()]} | {:error, term()}
  def vector_add(a, b) when length(a) == length(b) do
    result = Enum.zip_with(a, b, fn x, y -> x + y end)
    {:ok, result}
  end

  def vector_add(_a, _b) do
    {:error, :dimension_mismatch}
  end

  @doc """
  Subtract two vectors.
  """
  @spec vector_subtract([float()], [float()]) :: {:ok, [float()]} | {:error, term()}
  def vector_subtract(a, b) when length(a) == length(b) do
    result = Enum.zip_with(a, b, fn x, y -> x - y end)
    {:ok, result}
  end

  def vector_subtract(_a, _b) do
    {:error, :dimension_mismatch}
  end

  @doc """
  Multiply vector by scalar.
  """
  @spec vector_scale([float()], float()) :: [float()]
  def vector_scale(vector, scalar) do
    Enum.map(vector, &(&1 * scalar))
  end

  @doc """
  Calculate dot product of two vectors.

  ## Examples

      iex> UMICP.Matrix.Operations.dot_product([1.0, 2.0, 3.0], [4.0, 5.0, 6.0])
      {:ok, 32.0}
  """
  @spec dot_product([float()], [float()]) :: {:ok, float()} | {:error, term()}
  def dot_product(a, b) when length(a) == length(b) do
    result =
      Enum.zip_with(a, b, fn x, y -> x * y end)
      |> Enum.sum()

    {:ok, result}
  end

  def dot_product(_a, _b) do
    {:error, :dimension_mismatch}
  end

  @doc """
  Multiply two matrices.

  ## Parameters
  - `a`: First matrix (rows x cols) as flat list
  - `b`: Second matrix (cols x k) as flat list
  - `rows`: Number of rows in matrix A
  - `cols`: Number of columns in A / rows in B
  - `k`: Number of columns in matrix B

  ## Examples

      iex> UMICP.Matrix.Operations.matrix_multiply(
      ...>   [1.0, 2.0, 3.0, 4.0],
      ...>   [5.0, 6.0, 7.0, 8.0],
      ...>   2, 2, 2
      ...> )
      {:ok, [19.0, 22.0, 43.0, 50.0]}
  """
  @spec matrix_multiply([float()], [float()], pos_integer(), pos_integer(), pos_integer()) ::
          {:ok, [float()]} | {:error, term()}
  def matrix_multiply(a, b, rows, cols, k)
      when length(a) == rows * cols and length(b) == cols * k do
    result =
      for i <- 0..(rows - 1) do
        for j <- 0..(k - 1) do
          sum =
            for m <- 0..(cols - 1), reduce: 0.0 do
              acc ->
                a_val = Enum.at(a, i * cols + m)
                b_val = Enum.at(b, m * k + j)
                acc + a_val * b_val
            end

          sum
        end
      end
      |> List.flatten()

    {:ok, result}
  end

  def matrix_multiply(_a, _b, _rows, _cols, _k) do
    {:error, :invalid_dimensions}
  end

  @doc """
  Calculate vector magnitude (Euclidean norm).
  """
  @spec vector_magnitude([float()]) :: float()
  def vector_magnitude(vector) do
    vector
    |> Enum.map(&(&1 * &1))
    |> Enum.sum()
    |> :math.sqrt()
  end

  @doc """
  Normalize vector to unit length.
  """
  @spec vector_normalize([float()]) :: {:ok, [float()]} | {:error, term()}
  def vector_normalize(vector) do
    magnitude = vector_magnitude(vector)

    if magnitude > 0 do
      {:ok, vector_scale(vector, 1.0 / magnitude)}
    else
      {:error, :zero_vector}
    end
  end
end

File: bindings/go/pkg/umicp/matrix/operations.go
package matrix

import (
	"fmt"
	"math"
)

// DotProduct computes the dot product of two vectors
func DotProduct(a, b []float32) (float32, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("vectors must have the same length: got %d and %d", len(a), len(b))
	}

	var sum float32
	for i := range a {
		sum += a[i] * b[i]
	}

	return sum, nil
}

// Magnitude computes the Euclidean norm (magnitude) of a vector
func Magnitude(v []float32) float32 {
	var sum float32
	for _, val := range v {
		sum += val * val
	}
	return float32(math.Sqrt(float64(sum)))
}

// CosineSimilarity computes the cosine similarity between two vectors
func CosineSimilarity(a, b []float32) (float32, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("vectors must have the same length: got %d and %d", len(a), len(b))
	}

	dot, err := DotProduct(a, b)
	if err != nil {
		return 0, err
	}

	magA := Magnitude(a)
	magB := Magnitude(b)

	if magA == 0 || magB == 0 {
		return 0, nil
	}

	return dot / (magA * magB), nil
}

// EuclideanDistance computes the Euclidean distance between two vectors
func EuclideanDistance(a, b []float32) (float32, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("vectors must have the same length: got %d and %d", len(a), len(b))
	}

	var sum float32
	for i := range a {
		diff := a[i] - b[i]
		sum += diff * diff
	}

	return float32(math.Sqrt(float64(sum))), nil
}

// VectorAdd performs element-wise addition of two vectors
func VectorAdd(a, b, result []float32) error {
	if len(a) != len(b) || len(a) != len(result) {
		return fmt.Errorf("vectors must have the same length")
	}

	for i := range a {
		result[i] = a[i] + b[i]
	}

	return nil
}

// VectorSubtract performs element-wise subtraction of two vectors
func VectorSubtract(a, b, result []float32) error {
	if len(a) != len(b) || len(a) != len(result) {
		return fmt.Errorf("vectors must have the same length")
	}

	for i := range a {
		result[i] = a[i] - b[i]
	}

	return nil
}

// ScalarMultiply multiplies a vector by a scalar
func ScalarMultiply(v []float32, scalar float32, result []float32) error {
	if len(v) != len(result) {
		return fmt.Errorf("vectors must have the same length")
	}

	for i := range v {
		result[i] = v[i] * scalar
	}

	return nil
}

// MatrixMultiply multiplies two matrices
// A is m x n, B is n x p, result is m x p
func MatrixMultiply(a, b, result []float32, m, n, p int) error {
	if len(a) != m*n {
		return fmt.Errorf("matrix A size mismatch: expected %d, got %d", m*n, len(a))
	}
	if len(b) != n*p {
		return fmt.Errorf("matrix B size mismatch: expected %d, got %d", n*p, len(b))
	}
	if len(result) != m*p {
		return fmt.Errorf("result matrix size mismatch: expected %d, got %d", m*p, len(result))
	}

	for i := 0; i < m; i++ {
		for j := 0; j < p; j++ {
			var sum float32
			for k := 0; k < n; k++ {
				sum += a[i*n+k] * b[k*p+j]
			}
			result[i*p+j] = sum
		}
	}

	return nil
}

// Normalize normalizes a vector to unit length
func Normalize(v, result []float32) error {
	if len(v) != len(result) {
		return fmt.Errorf("vectors must have the same length")
	}

	mag := Magnitude(v)
	if mag == 0 {
		return fmt.Errorf("cannot normalize zero vector")
	}

	for i := range v {
		result[i] = v[i] / mag
	}

	return nil
}

// ManhattanDistance computes the Manhattan (L1) distance between two vectors
func ManhattanDistance(a, b []float32) (float32, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("vectors must have the same length: got %d and %d", len(a), len(b))
	}

	var sum float32
	for i := range a {
		diff := a[i] - b[i]
		if diff < 0 {
			diff = -diff
		}
		sum += diff
	}

	return sum, nil
}
